File: basic_learns/CPP_basic/OOP/OOP11/Account.cpp
#include "BankingCommonDecl.h"
#include "Account.h"
#include "AccountException.h"

Account::Account(int ID, int money, String name)
	: accID(ID), balance(money)
{
	cusName = name;
}

int Account::GetAccID() const { return accID; }

void Account::Deposit(int money) throw(InputException)
{
	if (money < 0)
	{
		throw InputException(money);			
	}
	balance += money;
}

int Account::Withdraw(int money)
{
		if (money < 0)
			throw InputException(money);
		if (balance < money)
			throw WithdrawException(balance);
		balance -= money;
		return money;	
}

void Account::ShowAccInfo() const
{
	cout << "°èÁÂID: " << accID << endl;
	cout << "ÀÌ  ¸§: " << cusName << endl;
	cout << "ÀÜ  ¾×: " << balance << endl;
}

File: basic_learns/java_basic/개인OOP/미완성/BankingSystem/src/Account.java
public class Account {
	private int accID;
	private String name;
	private int balance;

	public Account(int accID, String name, int balance) {
		this.accID = accID;
		this.name = name;
		this.balance = balance;
	}

	public void depositMoney(int money) throws MinusException {
		if(money<0) {
			throw new MinusException(money);
		}
		balance += money;
	}

	public int withdrawMoney(int money) 
			throws MinusException, InsuffException {
		if(money<0) {
			throw new MinusException(money);
		} else if(balance<money) {
			throw new InsuffException(money);
		}		
		balance -= money;
		return money;
	}

	public int getAccId() {
		return accID;
	}

	public void showAccInfo() {
		System.out.println("계좌 ID : " + accID);
		System.out.println("이름 : " + name);
		System.out.println("잔액 : " + balance);
	}
}

class NormalAccount extends Account {
	private int interest;

	NormalAccount(int accID, String name, int balance, int rate) {
		super(accID, name, balance);
		interest = rate;
	}

	public void depositMoney(int money) throws MinusException {
		super.depositMoney(money);
		super.depositMoney(money * interest);
	}
	
	public void showAccInfo() {
		super.showAccInfo();
		System.out.println("이자율 : "+interest);
	}
}

class CreditAccount extends NormalAccount {
	private int specialRate;
	
	public CreditAccount(int accID,String name, int balance, int rate, int specialRate) {
		super(accID,name,balance,rate);
		this.specialRate=specialRate;
	}
	
	public void depositMoney(int money) throws MinusException {
		super.depositMoney(money);
		(super.super).depositMoney(money*specialRate));
	}
	
	public void showAccInfo() {
		super.showAccInfo();
		System.out.println("특별 이자 : "+specialRate);
	}
	
}




















